src/Functions/FunctionsExternalDictionaries.h

```diff
@@ -140,6 +149,7 @@ class FunctionDictHelper
     const ExternalDictionariesLoader & external_loader;
     /// Access cannot be not granted, since in this case checkAccess() will throw and access_checked will not be updated.
     std::atomic<bool> access_checked = false;
+    std::unique_ptr<DictionaryStructure> structure_holder;
```

Comment:
It seems like there can be a race condition, because this helper is used in IFunction getReturnType, and this getReturnType can be called from multiple threads.

Reply:
Yes, you're right. Actually we can't use that kind of caching here due to https://github.com/ClickHouse/ClickHouse/pull/16205. I've changed the code.